pkg/healthcheck/healthcheck.go

```diff
@@ -2191,6 +2210,83 @@ func checkResources(resourceName string, objects []runtime.Object, expectedNames
 	return nil
 }
 
+// Check if there's a pod with the "opaque ports" annotation defined but a
+// service selecting the aforementioned pod doesn't define it
+func (hc *HealthChecker) checkMisconfiguredOpaquePortAnnotations(ctx context.Context, namespace string) error {
+	services, err := hc.GetServices(ctx)
+	if err != nil {
+		return err
+	}
+
+	var errStrings []string
+
+	for _, service := range services {
+		if service.Spec.ClusterIP == "None" {
+			// skip headless services; they're handled differently
+			continue
+		}
+
+		labelSet := labels.Set(service.Spec.Selector)
+		pods, err := hc.kubeAPI.CoreV1().Pods(namespace).List(
+			ctx,
+			metav1.ListOptions{LabelSelector: labelSet.AsSelector().String()},
+		)
+		if err != nil {
+			return err
+		}
+
+		if mismatch := misconfiguredOpaquePortAnnotationsInService(service, pods); mismatch != nil {
+			errStrings = append(
+				errStrings,
+				fmt.Sprintf("%s annotation is not properly configured in service %s",
+					k8s.ProxyOpaquePortsAnnotation,
+					service.Name),
+			)
+		}
+	}
+
+	if len(errStrings) >= 1 {
+		return fmt.Errorf(strings.Join(errStrings, "\n"))
+	}
+
+	return nil
+}
+
+func misconfiguredOpaquePortAnnotationsInService(service corev1.Service, pods *corev1.PodList) error {
+	for _, pod := range pods.Items {
+		if err := misconfiguredOpaqueAnnotation(service.Annotations, pod.Annotations); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func misconfiguredOpaqueAnnotation(svcAnnotations map[string]string, podAnnotations map[string]string) error {
+	// If the pod has the annotation, check that the service has it as well
+	if podAnnotation, found := podAnnotations[k8s.ProxyOpaquePortsAnnotation]; found {
+		if svcAnnotation, found := svcAnnotations[k8s.ProxyOpaquePortsAnnotation]; found {
+			if svcAnnotation != podAnnotation {
+				return fmt.Errorf("Both the pod and the service have the annotation but values don't match")
```

Comment:
To provide more information in these messages, it'd be helpful to provide names so that the user knows where to look! The function will need to accept the service and pods as a whole though so that we can get the name.

`fmt.Errorf("service and pod annotations do not match; service: %s, pod %s",  service.Name, pod.Name)`

And for when one is missing:

`fmt.Errorf("pod has annotation but service does not; service: %s, pod %s", service.Name, pod.Name)`

or

`fmt.Errorf("service has annotation but pod does not; service: %s, pod: %s",  service.Name, pod.Name)`

Reply:
Tried to improve the messaging here https://github.com/linkerd/linkerd2/pull/6192/commits/5bce6390fa41986277e47a4355e4e3fc929090f4